src/executor/graph_executor.cc

```diff
@@ -162,8 +171,9 @@ nnvm::Graph GraphExecutor::InitFullGraph(
   }
   if (!need_grad) return g;
   for (size_t i = 0; i < g.outputs.size(); ++i) {
-    head_grad_entry_.emplace_back(NodeEntry{nnvm::Node::Create(), 0, 0});
-    head_grad_map_[head_grad_entry_.back().node.get()] = i;
+    NodeEntry ngrad{nnvm::Node::Create(), 0, 0};
+    head_grad_entry_.emplace_back(AttrHint(ngrad, g.outputs[i]));
```

Comment:
This is for enforcing shape constraints in the gradient input. This is needed, in case some of the backward node is simply forward path, Say gradient of add returns an identity of ograd, and we need shape of the gradient input variable to start with